server/persistence/dynamo/dynamo_service_interaction.go

```diff
@@ -0,0 +1,244 @@
+	return is.write(writePlayerMeansOptions{
+		pm:         pm,
+		isCreation: true,
+	})
+}
+
+func (is *interactionService) Update(pm interaction.PlayerMeans) error {
+	return is.write(writePlayerMeansOptions{
+		pm: pm,
+	})
+}
+
+type writePlayerMeansOptions struct {
+	pm         interaction.PlayerMeans
+	isCreation bool
+}
+
+func (is *interactionService) write(opts writePlayerMeansOptions) error {
+	data := map[string]types.AttributeValue{
+		partitionKey: &types.AttributeValueMemberS{
+			Value: string(opts.pm.PlayerID),
+		},
+		sortKey: &types.AttributeValueMemberS{
+			Value: getSortKeyPrefix(is),
+		},
+		is.getPreferKey(): &types.AttributeValueMemberN{
+			Value: strconv.Itoa(int(opts.pm.PreferredMode)),
+		},
+	}
+
+	pii := &dynamodb.PutItemInput{
+		TableName: aws.String(dbName),
+		Item:      data,
+	}
+
+	if opts.isCreation {
+		// Use a conditional expression to only write items if this
+		// <HASH:RANGE> tuple doesn't already exist.
+		// See: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html
+		// and https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.OperatorsAndFunctions.html
+		pii.ConditionExpression = getConditionExpression(notExists, ``, notExists, ``)
+	}
+
+	_, err := is.svc.PutItem(is.ctx, pii)
+	if err != nil {
+		if isConditionalError(err) {
+			return persistence.ErrInteractionAlreadyExists
+		}
+		return err
+	}
+
+	for _, m := range opts.pm.Interactions {
+		pii, err = is.getPutItemInputForMeans(opts.pm.PlayerID, m)
+		if err != nil {
+			return err
+		}
+		_, err = is.svc.PutItem(is.ctx, pii)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (is *interactionService) getPutItemInputForMeans(
+	playerID model.PlayerID,
+	m interaction.Means,
+) (*dynamodb.PutItemInput, error) {
+	info, err := m.GetSerializedInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dynamodb.PutItemInput{
+		TableName: aws.String(dbName),
+		Item: map[string]types.AttributeValue{
+			partitionKey: &types.AttributeValueMemberS{
+				Value: string(playerID),
+			},
+			sortKey: &types.AttributeValueMemberS{
+				Value: is.getSpecForInteractionMeans(m),
+			},
+			is.getInfoKey(): &types.AttributeValueMemberB{
+				Value: info,
+			},
+		},
+	}, nil
+}
+
+func (is *interactionService) getSpecForInteractionMeans(
+	m interaction.Means,
+) string {
+	return getSortKeyPrefix(is) + `|` + strconv.Itoa(int(m.Mode))
+}
+
+func (is *interactionService) getInteractionMeansModeFromSpec(s string) (interaction.Mode, error) {
+	s = strings.TrimPrefix(s, getSortKeyPrefix(is)+`|`)
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		return interaction.Unknown, err
+	}
+	return interaction.Mode(i), nil
+}
+
+func (is *interactionService) getInfoKey() string {
+	return interactionInfoAttributeName
+}
+
+func (is *interactionService) getPreferKey() string {
+	return interactionPreferAttributeName
+}
```

Comment:
Why do these benefit from being functions? At the worst we get hit with some function all overhead (okay I'm not _actually_ worried about that), but at best it's redundant

Reply:
Ope, saw that you explicitly say you'd have them live in a `func`. Question still stands, though, why?

Reply:
Fair enough -- we'll save that for a rainy day :shrug: